syft/frameworks/torch/crypto/securenn.py

```diff
@@ -121,96 +121,92 @@ def select_share(alpha_sh, x_sh, y_sh):
-    R_MASK = R_MASK.unsqueeze(1).expand(list(r.shape))
+    j = sy.MultiPointerTensor(children=[torch.tensor([0]).send(alice), torch.tensor([1]).send(bob)])
 
-    u = (torch.rand(x.shape) > 0.5).long().send(bob, alice).child
-    # Mask for condition i̸=1 in 11)
-    l1_mask = torch.zeros(x.shape).long()
-    l1_mask[:, -1:] = 1
-    l1_mask = l1_mask.send(bob, alice).child
+    # 1)
+    t = (r + 1) % L
+    t_bit = decompose(t)
+    r_bit = decompose(r)
 
-    # if BETA == 0
+    # if beta == 0
     # 5)
-    j0 = torch.zeros(x.shape).long().send(bob)
-    j1 = torch.ones(x.shape).long().send(alice)
-    j = sy.MultiPointerTensor(children=[j0, j1])
-    w = (j * r) + x - (2 * x * r)
-
+    w = x_bit_sh + (j * r_bit) - (2 * r_bit * x_bit_sh)
     # 6)
-    wf = flip(w, 1)
-    wfc = wf.cumsum(1) - wf
-    wfcf = flip(wfc, 1)
-    c_beta0 = (j * r) - x + j + wfcf
+    wc = w.flip(-1).cumsum(-1).flip(-1) - w
+    c_beta0 = -x_bit_sh + (j * r_bit) + j + wc
 
-    # elif BETA == 1 AND r != 2^l- 1
+    # elif beta == 1 AND r != 2^l- 1
     # 8)
-    w = x + (j * t) - (2 * t * x)  # FIXME: unused
+    w = x_bit_sh + (j * t_bit) - (2 * t_bit * x_bit_sh)
     # 9)
-    c_beta1 = (-j * t) + x + j + wfcf
+    wc = w.flip(-1).cumsum(-1).flip(-1) - w
+    c_beta1 = x_bit_sh + (-j * t_bit) + j + wc
 
     # else
     # 11)
     c_igt1 = (1 - j) * (u + 1) - (j * u)
-    c_ie1 = (j * -2) + 1
-    c_21l = (l1_mask * c_ie1) + ((1 - l1_mask) * c_igt1)
+    c_ie1 = (1 - 2 * j) * u
+
+    l1_mask = torch.zeros(x_bit_sh.shape).long()
+    l1_mask[..., 0] = 1
+    l1_mask = l1_mask.send(alice, bob).child
+    # c_else = if i == 1 c_ie1 else c_igt1
+    c_else = (l1_mask * c_ie1) + ((1 - l1_mask) * c_igt1)
+
+    # Mask for the case r == 2^l −1
+    r_mask = (r == (L - 1)).long()
+    r_mask = r_mask.unsqueeze(-1)
 
     # Mask combination to execute the if / else statements of 4), 7), 10)
-    c = (1 - BETA) * c_beta0 + BETA * c_beta1
-    c = (c * (1 - R_MASK)) + (c_21l * R_MASK)
+    c = (1 - beta) * c_beta0 + (beta * (1 - r_mask)) * c_beta1 + (beta * r_mask) * c_else
 
     # 14)
     # Hide c values
-    s = torch.randint(1, p, c.shape).send(alice, bob).child
     mask = s * c
+
     # Permute the mask
-    perm = torch.randperm(c.shape[1]).send(alice, bob).child
-    permuted_mask = mask[:, perm]
-    # Send it to another worker
-    remote_mask = permuted_mask.wrap().send(crypto_provider)
+    # I have to create idx because Ellipsis are still not supported
+    # (I would like to do permuted_mask = mask[..., perm])
+    idx = tuple([slice(None)] * (len(x_bit_sh.shape) - 1) + [perm])
+    permuted_mask = mask[idx]
 
     # 15)
-    # transform remotely the AdditiveSharingTensor back to a torch tensor
-    d_ptr = remote_mask.remote_get()
-    beta_prime = (d_ptr == 0).sum(1)
-    # Get result back
-    res = beta_prime.get()
-    return res
+    d = permuted_mask.get()
+    beta_prime = ((d == 0).sum(-1) > 0).long()
```

Comment:
The sum should be able to exceed 1 right? It should be only 0 or 1

Reply:
I think you're right!